pkg/ebpf/uprobes/inspector.go

```diff
@@ -0,0 +1,176 @@
+			bestEffortSymbols.Add(req.Name)
+		} else {
+			mandatorySymbols.Add(req.Name)
+		}
+
+		if req.IncludeReturnLocations {
+			return nil, false, fmt.Errorf("return locations are not supported by the native binary inspector")
+		}
+	}
+
+	symbolMap, err := bininspect.GetAllSymbolsByName(elfFile, mandatorySymbols)
+	if err != nil {
+		return nil, false, err
+	}
+	/* Best effort to resolve symbols, so we don't care about the error */
+	symbolMapBestEffort, _ := bininspect.GetAllSymbolsByName(elfFile, bestEffortSymbols)
+
+	funcMap := make(map[string]bininspect.FunctionMetadata, len(symbolMap)+len(symbolMapBestEffort))
+	for symbol, sym := range symbolMap {
+		m, err := p.symbolToFuncMetadata(elfFile, sym)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to convert symbol %v to function metadata: %w", sym, err)
+		}
+		funcMap[symbol] = *m
+	}
+	for symbol, sym := range symbolMapBestEffort {
+		m, err := p.symbolToFuncMetadata(elfFile, sym)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to convert symbol %v to function metadata: %w", sym, err)
+		}
+		funcMap[symbol] = *m
+	}
+
+	return funcMap, true, nil
+}
+
+func (*NativeBinaryInspector) symbolToFuncMetadata(elfFile *elf.File, sym elf.Symbol) (*bininspect.FunctionMetadata, error) {
+	manager.SanitizeUprobeAddresses(elfFile, []elf.Symbol{sym})
+	offset, err := bininspect.SymbolToOffset(elfFile, sym)
+	if err != nil {
+		return nil, err
+	}
+
+	return &bininspect.FunctionMetadata{EntryLocation: uint64(offset)}, nil
+}
+
+func (*NativeBinaryInspector) Cleanup(_ utils.FilePath) {
+	// Nothing to do here for the native inspector
+}
+
+// GoBinaryInspector is a BinaryInspector that inspects Go binaries, dealing with the specifics of Go binaries
+// such as the argument passing convention and the lack of uprobes
+type GoBinaryInspector struct {
+	structFieldsLookupFunctions map[bininspect.FieldIdentifier]bininspect.StructLookupFunction
+	paramLookupFunctions        map[string]bininspect.ParameterLookupFunction
+}
+
+// Ensure GoBinaryInspector implements BinaryInspector
+var _ BinaryInspector = &GoBinaryInspector{}
+
+// Inspect extracts the metadata required to attach to a Go binary from the ELF file at the given path.
+func (p *GoBinaryInspector) Inspect(fpath utils.FilePath, requests []SymbolRequest) (map[string]bininspect.FunctionMetadata, bool, error) {
+	path := fpath.HostPath
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not open file %s, %w", path, err)
+	}
+	defer f.Close()
+
+	elfFile, err := elf.NewFile(f)
+	if err != nil {
+		return nil, false, fmt.Errorf("file %s could not be parsed as an ELF file: %w", path, err)
+	}
+
+	functionsConfig := make(map[string]bininspect.FunctionConfiguration, len(requests))
+	for _, req := range requests {
+		functionsConfig[req.Name] = bininspect.FunctionConfiguration{
+			IncludeReturnLocations: req.IncludeReturnLocations,
+			ParamLookupFunction:    p.paramLookupFunctions[req.Name],
```

Comment:
I think it will be great to validate `req.Name` exists in `p.paramLookupFunctions`, and to fail if it does not